docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,544 @@
+1. As a cluster operator, I want to know if my cluster’s control plane is healthy in order to understand if I am meeting my SLOs with my end users.
+1. As a cluster operator, I want to be able to quickly respond to a Kubernetes CVE by upgrading my clusters in an automated fashion.
+1. As a cluster operator, I want to be able to quickly respond to a non-Kubernetes CVE that affects my base image or Kubernetes dependencies by upgrading my clusters in an automated fashion.
+1. As a cluster operator, I would like to upgrade to a new minor version of Kubernetes so that my cluster remains supported.
+1. As a cluster operator, I want to know why my cluster isn’t working properly after creation. I have ended up with an API server I can access, but CNI isn’t working, and new machines are not registering themselves with the control plane.
+
+#### Identified features from user stories
+
+1. Based on the function of kubeadm, the control plane provider must be able to scale the number of replicas of a control plane from 1 to X, meeting user stories 1 through 4.
+1. To address user story 5, the control plane provider must provide validation of the number of replicas in a control plane. Where the stacked etcd topology is used (i.e., in the default implementation), the number of replicas must be an odd number, as per [etcd best practice](https://etcd.io/docs/v3.3.12/faq/#why-an-odd-number-of-cluster-members). When external etcd is used, any number is valid.
+1. In service of user story 5, the control plane provider must also manage etcd membership via kubeadm as part of scaling down (`kubeadm` takes care of adding the new etcd member when joining).
+1. The control plane provider should provide indicators of health to meet user story 6 and 10. This should include at least the state of etcd and information about which replicas are currently healthy or not. For the default implementation, health attributes based on artifacts kubeadm installs on the cluster may also be of interest to cluster operators.
+1. The control plane provider must be able to upgrade a control plane’s version of Kubernetes as well as updating the underlying machine image on where applicable (e.g. virtual machine based infrastructure).
+
+### Implementation Details/Notes/Constraints
+
+#### New API Types
+
+ControlPlane:
+
+```go
+package v1alpha3
+
+import (
+    corev1  "k8s.io/api/core/v1"
+    metav1  "k8s.io/apimachinery/pkg/apis/meta/v1"
+    cabpkv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+)
+
+// ControlPlaneSpec defines the desired state of ControlPlane
+type ControlPlaneSpec struct {
+    // Number of desired machines. Defaults to 1.
+    // This is a pointer to distinguish between explicit zero and not specified.
+    Replicas *int32 `json:"replicas"`
```

Comment:
> I'm very concerned about a situation of how a control plane with static IPs is supposed to exist with the control plane management taking on a model similar to a MachineDeployment with one or more replicas.

Is this something that you believe needs to be solved by the initial implementation? Or is it something that we can likely add through iteration later?

Reply:
Hi @detiber,

> Is this something that you believe needs to be solved by the initial implementation? Or is it something that we can likely add through iteration later?

This is a case where I think it's important to call out how this approach works with on-prem environments that don't have the vertical solution for compute instance/network management like AWS does for example. It's already an issue for worker nodes -- no good way to assign static IPs via MachineDeployments -- and this new model just extends the issue to the control plane.

Additionally, consider load balancers. CAPA has a cheat for LBs - it doesn't need to know the IPs of instances to connect an LB's target groups to specific instances, AWS ELBs can connect to instance refs before ever knowing the IP of an instance.

For CAPV we must know the IPs to connect LBs with the tech currently available. And this is likely true for other on-prem providers that exist today or might exist. When it comes to bringing your own LB ahead of time, this is possible because we can simply pre-define the LB's targets with static IPs and ensure the control plane machines are configured to use these static IPs. With this new control plane model, we'd be limited to DHCP addressing, making it nearly impossible to connect an existing load balancer with any type of acceptable UX.

Reply:
@akutz Thanks for bringing this up. I agree strongly that on-prem is an important use case, and that we should enable it as much as possible.

As for updating an external LB, I think there are a couple options that work with the current proposal . 
1. Have each Machine update the LB during bootstrapping; that's possible by adding pre/post commands to the KubeadmConfig.
2. Preallocate a pool of static IPs and configure the LB with them. The Infra Provider can follow label/annotation on a Machine to request an instance with an IP from the preallocated pool.

Sometimes an external LB is not available. Then a VIP might be used. That's a use case that should, I think, work with the default implementation, but maybe not in this first proposal.

Reply:
@thebsdbox any thoughts from your experimentations?

Reply:
I still want to be able to differentiate between nil and 0, since users might be confused if they explicitly set it to 0 and it magically becomes 1, vs nil becoming 1.

Reply:
Load balancer (or some other externally stable endpoint) has now been listed as a requirement.